pkg/k8s/apis/cilium.io/v2alpha1/bgp_peer_types.go

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package v2alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=false
+// +deepequal-gen=false
+
+// CiliumBGPPeerTemplateList is a list of CiliumBGPPeer objects.
+type CiliumBGPPeerTemplateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// Items is a list of CiliumBGPPeer.
+	Items []CiliumBGPPeerTemplate `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:categories={cilium,ciliumbgp},singular="ciliumbgppeertemplate",path="ciliumbgppeertemplates",scope="Cluster",shortName={bgppeer}
+// +kubebuilder:printcolumn:JSONPath=".metadata.creationTimestamp",name="Age",type=date
+// +kubebuilder:storageversion
+
+type CiliumBGPPeerTemplate struct {
```

Comment:
As I mentioned in my other comment, IMHO we should consider adding `peerGroups` to the CiliumBGPClusterConfig CRD. A `peerGroup` would define BGP config that applies to all the peers in the group. For example:

```yaml
apiVersion: cilium.io/v1alpha1
kind: CiliumBGPClusterConfig
metadata:
  name: example-config
spec:
  nodeSelector:
    matchLabels:
      rack: rack1
  bgpInstances:
  - localASN: 65001
    peerGroups:
    - name: example-group
      peers:
      - name: peer1
        address: 10.0.0.1
      - name: peer2
        address: 10.0.0.2
      - name: peer3
        address: 10.0.0.3
      peerASN: 65002
      eBGPMultihopTTL: 3
     ...
```

WDYT?

Reply:
Not sure how this is removing duplication of configuration among multiple CiliumBGPClusterConfig. 

```yaml
apiVersion: cilium.io/v1alpha1
kind: CiliumBGPClusterConfig
metadata:
  name: example-config
spec:
  nodeSelector:
    matchLabels:
      rack: rack1
  bgpInstances:
  - localASN: 65001
    peerGroups:
    - name: example-group
      peers:
      - name: peer1
        address: 10.0.0.1
      peerASN: 65002
      eBGPMultihopTTL: 3
     ... common config ...
```

For rack 2, we will have this configuration

```
apiVersion: cilium.io/v1alpha1
kind: CiliumBGPClusterConfig
metadata:
  name: example-config-2
spec:
  nodeSelector:
    matchLabels:
      rack: rack2
  bgpInstances:
  - localASN: 65001
    peerGroups:
    - name: example-group-2
      peers:
      - name: peer2
        address: 10.0.0.2
      peerASN: 65002
      eBGPMultihopTTL: 3
     ... common config ...
```

Node selector associates example-config to rack 1 and example-config-2 to rack 2. 

Common configuration is repeated between example-config and example-config-2. This common configuration can be lot of BGP knobs which is eliminated if we use peer template.

Reply:
Based on ^ use case, I can see why the template-based approach is preferred. Thanks for the additional details.

Reply:
@danehans  Let me know if we can resolve this comment. Let's discuss design in CFP, and whether we need to modify this approach or not.

Reply:
Yes, this comment can be resolved and any additional discussion on this topic can take place in the design doc.